### ISSUE TITLE
Open-ended VQA-RAD evaluation always returns 0 while close-ended matches the paper

### ISSUE THREAD

**Issue by folgemir0630** [GaryGuTC/LaPA_model#13]
Hi authors,

We carefully reproduced LaPA following the exact software environment specified in the README:

- PyTorch: 1.9.0
- Transformers: 4.6.0
- PyTorch Lightning: 1.3.2
- GPU: NVIDIA A100 (Ampere)
- Tokenizer: RobertaTokenizerFast with tokenizer.json properly generated

The training process runs normally and the loss decreases as expected.
We evaluated the official checkpoint on VQA-RAD using test.sh.

Our test results are:

- Close-ended accuracy: ~0.82 (very close to the reported baseline)
- Open-ended accuracy: 0.0
- Overall score: ~0.45

This behavior is consistent across multiple runs and seeds.
Importantly, the close-ended performance matches the paper, while the open-ended score is always zero.

This suggests that:
- The open-ended evaluation path may rely on undocumented preprocessing, tokenizer assumptions, or span-alignment logic.
- Or there may be a bug or missing step in the released evaluation code for open-ended VQA-RAD.

Given that close-ended evaluation works correctly under the same setting, we believe this is not an environment or training issue.

Could you please clarify:
1. Whether any additional preprocessing or filtering is required for open-ended VQA-RAD answers?
2. Whether a specific tokenizer configuration or post-processing step is assumed but not documented?
3. Whether the released open-ended evaluation code reflects the one used to report the paper results?

Thank you for your time, and we hope this report can help improve reproducibility for future users.

<img width="411" height="144" alt="Image" src="https://github.com/user-attachments/assets/2a65f5e8-c3ba-401d-9475-259db497c46c" />

As shown in the attached screenshot, when evaluating the official VQA-RAD checkpoint using the provided test script and the README-specified environment, we consistently observe:

- Close-ended accuracy around 0.82, which closely matches the reported baseline.
- Open-ended accuracy equal to 0.0 across runs.

This clear discrepancy suggests that the open-ended evaluation path behaves fundamentally differently from the close-ended one, despite using the same model, tokenizer, and environment.

Since the close-ended results are reproducible and consistent with the paper, this issue is unlikely to be caused by environment mismatch or training instability. Instead, it indicates that the open-ended evaluation may rely on additional preprocessing steps, tokenizer assumptions, or post-processing logic that are not documented in the released code.

**Comment by GaryGuTC** [comment]
Hi @folgemir0630

Thank you for your attention. Here are the responses to your questions:

> 1. Whether any additional preprocessing or filtering is required for open-ended VQA-RAD answers?

No additional preprocessing is required. **You may double-check the evaluation dataset and the corresponding code.**

> 2. Whether a specific tokenizer configuration or post-processing step is assumed but not documented?

No, it is not. I have updated all related information, including the code and data.

> 3. Whether the released open-ended evaluation code reflects the one used to report the paper results?

Yes, it does. I have previously reproduced the results using this code.

**Comment by folgemir0630** [comment]
Hi authors,

thank you for your previous reply and for clarifying that no additional preprocessing or undocumented tokenizer assumptions are required, and that the released evaluation code reflects what was used in the paper.

After further investigation, we would like to share additional technical findings and ask for clarification regarding the open-ended VQA-RAD results reported in the paper.

1. Evaluation mechanism (code-level observation)

By tracing the evaluation pipeline, we confirmed that both close-ended and open-ended VQA-RAD samples are evaluated using the same argmax-based multi-class classification metric (VQARADScore in my_metrics.py). Specifically:
- The model output logits are reduced via argmax.
- The prediction is converted to a one-hot vector.
- The score is computed by checking whether the predicted class index belongs to the ground-truth multi-hot answer set.
There is no generative decoding, semantic matching, or string-level comparison in the open-ended evaluation path.

2. Top-k sanity check on open-ended samples

To verify whether the issue is caused by overly strict top-1 evaluation, we extended the metric to compute Top-k hit rates on open-ended samples (k = 5, 30), without changing the model or dataset.

Using the official VQA_RAD_best.ckpt and the released test script, we consistently observe:
- Open Top-1 accuracy: 0.0
- Open Top-5 accuracy: ≈ 0.0 (only extremely rare hits)
- Open Top-30 accuracy: ≈ 0.05–0.06

This indicates that, for most open-ended questions, the ground-truth answer labels are not ranked within the top 30 logits, suggesting that the issue is not caused by the strictness of top-1 evaluation, but by the logit ranking itself on open-ended samples.

3. Control experiment

As a control, when training the model ourselves (e.g., with latent prompts enabled), we obtain non-zero open-ended scores (~0.2), confirming that:
- The evaluator itself can yield non-zero open-ended scores.
- The observed behavior is specific to the official checkpoint rather than a bug in the metric implementation.

4. Clarification request

Given the above, we would like to ask for clarification on the following point:

Under the current public code and evaluation definition (argmax-based classification on a fixed answer vocabulary), how were the open-ended VQA-RAD results reported in the paper obtained?

Specifically, were any of the following used in the paper experiments:
- A different answer vocabulary or label filtering strategy for open-ended questions?
- A restricted subset of open-ended answers (e.g., high-frequency answers only)?
- A different evaluation implementation or configuration that is not reflected in the current repository?

We believe clarifying this would greatly help the community correctly interpret the open-ended VQA-RAD results and improve reproducibility.

Thank you very much for your time and for making the code publicly available.

The modified top-5 statistics code is as follows:(/m3ae/gadgets/my_metrics.py)

`def update(self, logits, target, types=None):
    logits, target = (
        logits.detach().float().to(self.score.device),
        target.detach().float().to(self.score.device),
    )

    pred1 = torch.argmax(logits, dim=1)
    one_hot = torch.zeros_like(target)
    one_hot.scatter_(1, pred1.view(-1, 1), 1)
    scores = one_hot * target

    close_scores = scores[types == 0]
    open_scores  = scores[types == 1]

    self.close_score += close_scores.sum()
    self.close_total += len(close_scores)
    self.open_score  += open_scores.sum()
    self.open_total  += len(open_scores)

    top5 = torch.topk(logits, k=5, dim=1).indices   # [B, 5]
    open_mask = (types == 1)

    if open_mask.any():
        gt_open = target[open_mask]          # [N_open, V]
        top5_open = top5[open_mask]          # [N_open, 5]

        hit_top5 = []
        for i in range(top5_open.size(0)):
            hit = gt_open[i, top5_open[i]].sum() > 0
            hit_top5.append(hit)

        hit_top5 = torch.tensor(hit_top5, device=self.score.device, dtype=torch.float)
        self.open_score_top5 += hit_top5.sum()
        self.open_total_top5 += hit_top5.numel()

    self.score += scores.sum()
    self.total += len(scores)`

**Comment by GaryGuTC** [comment]
Hi @folgemir0630

Thank you for your detailed investigation. Regarding your questions:

1. As this work was completed some time ago, I no longer recall all the experimental details precisely, and the original code has unfortunately been lost. My apologies for this limitation.
2. I suspect the dataset you downloaded may not have been prepared correctly. Please follow the official dataset preparation guidelines:
   - LaPA dataset instructions: [link](https://github.com/GaryGuTC/LaPA_model?tab=readme-ov-file#dataset)
   - M3AE dataset instructions: [link](https://github.com/zhjohnchan/M3AE?tab=readme-ov-file#1-dataset-preparation-1)
3. Please note that our dataset and infrastructure were set up following the [M3AE repository](https://github.com/zhjohnchan/M3AE). If you encounter issues during evaluation, I recommend reviewing its GitHub issues for potential solutions. This specific discussion might be relevant: [link](https://github.com/zhjohnchan/M3AE/issues/3#issuecomment-1258985614).

Best regards,
Tiancheng

**Comment by folgemir0630** [comment]
Hi Tiancheng,

thank you very much for your reply and for the clarification.

After further investigation on our side, we have identified that the consistently zero open-ended score on VQA-RAD is mainly due to the interaction between the released evaluation protocol (argmax-based closed-set classification) and the highly long-tailed, single-label nature of open-ended answers in VQA-RAD, rather than an implementation or environment issue.

Given the lack of preserved experimental code and the difficulty of fully aligning the original experimental setting, we believe this point is now sufficiently understood, and we will close this issue.

Thank you again for making the code available and for your time.

Best regards,
folgemir
